src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -265,79 +273,189 @@ CreateReferenceTable(Oid relationId)
+	colocationId = CreateReferenceTableColocationId();
+	CreateDistributedTable(relationId, distributionColumn, DISTRIBUTE_BY_NONE,
+						   colocationId, viaDeprecatedAPI);
+
+	relation_close(relation, NoLock);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * CreateDistributedTable creates distributed table in the given configuration.
+ * It took care of locking, replication settings, colocation, metadata creation,
+ * creating shards and copying local data to shards.
+ */
+void
+CreateDistributedTable(Oid relationId, Var *distributionColumn, char distributionMethod,
+					   uint32 colocationId, bool viaDeprecatedAPI)
+{
+	char replicationModel = REPLICATION_MODEL_INVALID;
+	int replicationFactor = ShardReplicationFactor;
+	Oid colocatedTableId = InvalidOid;
+	bool localTableEmpty = false;
+
+	EnsureDistributionConfiguration(relationId, distributionColumn, distributionMethod,
+									colocationId, replicationModel, viaDeprecatedAPI);
+
+	/* we need to calculate these variables before creating distributed metadata */
+	localTableEmpty = LocalTableEmpty(relationId);
+	colocatedTableId = ColocatedTableId(colocationId);
+	if (colocatedTableId != InvalidOid)
 	{
-		EnsureTableNotDistributed(relationId);
+		replicationFactor = TableShardReplicationFactor(colocatedTableId);
+	}
+	replicationModel = AppropriateReplicationModel(distributionMethod, replicationFactor,
+												   viaDeprecatedAPI);
+
+	/* create an entry for distributed table in pg_dist_partition */
+	InsertIntoPgDistPartition(relationId, distributionMethod, distributionColumn,
+							  colocationId, replicationModel);
+
+	/* foreign tables does not support TRUNCATE trigger */
+	if (RegularTable(relationId))
+	{
+		CreateTruncateTrigger(relationId);
+	}
+
+	/* if we are using master_create_distributed_table, we don't need to continue */
+	if (viaDeprecatedAPI)
+	{
+		return;
+	}
+
+	/* create shards for hash distributed and reference tables */
+	if (distributionMethod == DISTRIBUTE_BY_HASH)
+	{
+		CreateHashDistributedTableShards(relationId, colocatedTableId, localTableEmpty);
+	}
+	else if (distributionMethod == DISTRIBUTE_BY_NONE)
+	{
+		CreateReferenceTableShard(relationId);
+	}
+
+	/* copy over data for hash distributed and reference tables */
+	if (distributionMethod == DISTRIBUTE_BY_HASH ||
+		distributionMethod == DISTRIBUTE_BY_NONE)
+	{
+		if (RegularTable(relationId))
+		{
+			CopyLocalDataIntoShards(relationId);
+		}
+	}
+
+	if (ShouldSyncTableMetadata(relationId))
+	{
+		CreateTableMetadataOnWorkers(relationId);
+	}
+}
+
+
+/*
+ * AppropriateReplicationModel function returns appropriate replication model
+ * depending on distributionMethod and global ReplicationModel variable
+ */
+static char
+AppropriateReplicationModel(char distributionMethod, int replicationFactor,
+							bool viaDeprecatedAPI)
+{
+	if (viaDeprecatedAPI)
+	{
+		if (ReplicationModel != REPLICATION_MODEL_COORDINATOR)
+		{
+			ereport(NOTICE, (errmsg("using statement-based replication"),
```

Comment:
We already don't support this feature, so I'd rather error out on this. @byucesoy could you please change this to `ERROR`?